### PR TITLE
Migrate AT_DISPATCH_COMPLEX_TYPES to c10::complex

### PR DESCRIPTION
**Pull request by zasdfgbnm**

### CODE REVIEW

File: aten/src/ATen/native/cuda/PowKernel.cu [hunk 1]
```diff
@@ -156,11 +156,18 @@ void pow_tensor_scalar_kernel_impl(TensorIterator& iter,
...
+#if defined(CUDA_VERSION) && (CUDA_VERSION < 10000)
+#define CUDA92_BUG(x) thrust::complex<T>(x.real(), x.imag())
+#else
+#define CUDA92_BUG(x) x
+#endif
+
 void pow_tensor_scalar_kernel(TensorIterator& iter, Scalar exp_scalar) {
   if (isComplexType(iter.dtype()) || exp_scalar.isComplex()) {
     AT_DISPATCH_COMPLEX_TYPES(iter.dtype(), "pow_cuda", [&]() {
       using thrust_t = typename ztype_cuda<scalar_t>::thrust_t;
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/PowKernel.cu:168]: It is not easy to migrate this from `thrust::complex` to `c10::complex` due to bugs on `std::pow` on Windows nvcc, and I don't want this PR to be blocked by this. So in this PR, I choose to keep `thrust::complex` and workaround the `c10::complex --> thrust::complex` cast.

File: aten/src/ATen/native/RangeFactories.cpp [hunk 2]
```diff
@@ -30,6 +30,16 @@ Tensor& linspace_cpu_out(Tensor& result, Scalar start, Scalar end, int64_t steps
   return result;
 }
 
+// c10::complex fails on MacOS with SIGILL (illegal instruction)
+// for unknown reason. See: https://github.com/pytorch/pytorch/issues/39123
+// I don't know what is the reason, but looks like a compiler bug
+//  -- @zasdfgbnm
+#ifdef __APPLE__
```
Comment (anjali411) [inline aten/src/ATen/native/RangeFactories.cpp:37]: nit `__MACOS_ERROR__`

**Comment by zasdfgbnm** [comment]
@anjali411 This should be ready as well

**Comment by zasdfgbnm** [comment]
I messed up this PR, closing and in favor of https://github.com/pytorch/pytorch/pull/39553

**Pull request closed without merging**